### PR TITLE
timeout threshold flag

### LINKED ISSUE #1431: Stop running rules on file that keeps timing out
With `--timeout` flag it is still possible for a file to cause a timeout on every single rule.

Make default behavior to be to stop trying to scan a file after X number of rule timeouts. Add a flag that modifies the value of X.

This will significantly speed up run times when running on a codebase with a problematic file.

### PR DESCRIPTION
**Pull request by sabrinabrogren**
Set `max-timeouts` to 3 for now, but open to suggestions on that (and everything else, including the flag name).

### CODE REVIEW

File: semgrep/tests/e2e/test_check.py
Comment (brendongo) [inline semgrep/tests/e2e/test_check.py]: Probably worth adding a test that has max-timeouts > 1 and checking that it doesn't stop running rules until it hits max-timeouts. Can probably just create another rule file that duplicates the rule in long.yaml a bunch of times.

Comment (sabrinabrogren) [reply semgrep/tests/e2e/test_check.py]: Will do!

**Comment by mschwager** [comment]
Had a more general question on Slack taking it back a step:

> Hmm, I wonder if it's even better to take a step back here. Would it make more sense to avoid adding another flag here and leverage something like `--exclude`? I.e. if we timeout more than once on the same file we print a message saying something like "Timing out on <filename>, consider excluding file from analysis with `--exclude`". The downside is they may have to restart their analysis, the upside seems to be avoiding a potentially confusing flag. As a `semgrep` user I'm not really sure when a file could timeout, and I don't really know a reasonable value to set here unless I need it to try more rules against a timing out file
>
> Or maybe the language is just a bit confusing to me. Something like `--timeout-retries` makes more sense to me, although that may imply you're retrying the same rule. Thoughts?

What do you think @brendongo?

File: semgrep/semgrep/core_runner.py
Comment (mschwager) [inline semgrep/semgrep/core_runner.py]: We can be slightly more defensive here by using `>=`, which could prevent an infinite loop in the future.

File: semgrep/semgrep/output.py
Comment (mschwager) [inline semgrep/semgrep/output.py:258]: Do we typically use `with_color` with our loggers? I see it's only currently used in `semgrep/error.py` and not in any `logger` calls. I'll defer to @brendongo here.

Comment (sabrinabrogren) [reply semgrep/semgrep/output.py:258]: I used it here because the messages were red before, so when aggregating them and creating a new error message I added `with-color` to make them look like the other not aggregated error messages.

Comment (brendongo) [reply semgrep/semgrep/output.py:258]: I think there are other errors that have color added as part of their __str__ function that get written to log so this is fine.

**Review by brendongo — APPROVED** [review]
Can you make the PR title more descriptive before merging

**Pull request merged**
